File: Ladder03/Python_code/binary_tree_preorder_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
"""
# non-recursive version
class Solution:
    """
    @param root: The root of binary tree.
    @return: Preorder in ArrayList which contains node values.
    """
    def preorderTraversal(self, root):
        if root is None:
            return []
        stack = []
        result = []
        while root or stack:
            if root:
                result.append(root.val)
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                root = root.right
        return result

    def preorderTraversal(self, root):
        if root == None:
            return []
        result = []
        stack = [root]
        while len(stack) != 0:
            node = stack.pop()
            result.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return result

# Divide and conquer
class Solution:
    def preorderTraversal(self, root):
        if not root:
            return []
        left = self.preorderTraversal(root.left)
        right = self.preorderTraversal(root.right)
        return [root.val] + left + right

    def preorderTraversal(self, root):
        return [root.val] + self.preorderTraversal(root.left) + self.preorderTraversal(root.right) if root else []

# recursive version
class Solution:
    def preorderTraversal(self, root):
        res = []
        if not root:
            return res
        self.dfs(root, res)
        return res

    def dfs(self, node, res):
        res.append(node.val)
        if node.left:
            self.dfs(node.left, res)
        if node.right:
            self.dfs(node.right, res)    

File: NineChapters/03.BinaryTree_DivideConquer/Java_code/BinaryTreePreorderTraversal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
// recursive version:
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in ArrayList which contains node values.
     */
    private ArrayList<Integer> result = new ArrayList<>();
    public ArrayList<Integer> preorderTraversal(TreeNode root) {
        if (root == null) {
            return result;
        }
        preorderTraversalHelper(root);
        return result;
    }

    private void preorderTraversalHelper(TreeNode root) {
        if (root == null) {
            return;
        }
        result.add(root.val);
        preorderTraversalHelper(root.left);
        preorderTraversalHelper(root.right);
    }
}
// non-recursive version:
public class Solution {
    public ArrayList<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            TreeNode node = stack.pop();
            result.add(node.val);
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
        }
        return result;
    }
}